sled-agent/config-reconciler/src/reconciler_task/external_disks.rs

```
@@ -158,6 +162,59 @@ impl CurrentlyManagedZpoolsReceiver {
             }
         }
     }
+
+    // This returns a tuple that can be converted into an `InventoryZpool`. It
+    // doesn't return an `InventoryZpool` directly because the latter only
+    // contains the zpool's ID, not the full name, and our caller wants the
+    // names too.
+    pub(crate) async fn to_inventory(
```

Is this the right name for this function?

most of the other `to_inventory` functions are not collecting additional data, they're doing a transformation synchronously. IMO the "to_" prefix implies "acts like `From/Into`".

This function appears to actually be executing ZFS commands, and silently dropping zpools that might be throwing an error (well, it's logging them, but they're getting filtered out of the results) .

- Should we make it possible to report "partial inventories" when a zpool/dataset is in a bad state, and we can't query it? E.g., should we report "this zpool exists, but we can't query info about it"?
- ... or are we okay with this current disposition, where we can see a zpool, we think it might exist, but we'll drop it from inventory entirely unless it responds "Ok" to enough requests?

---

This is lifted nearly verbatim from the current sled-agent `/inventory` handler: https://github.com/oxidecomputer/omicron/blob/c8d8f00a0b863e0bdada340e7dabdd84b6de4d13/sled-agent/src/sled_agent.rs#L1371-L1389

(It's slightly different in that a `ByteCount::try_from` failure here means "skip" and in sled-agent today means "fail the inventory collection altogether, but I think a `ByteCount::try_from` failure here is basically impossible in practice? Much more likely that the `Zpool::get_info()` fails, as you point out.)

Sorry to sound like an echo chamber from yesterday, but: I think your questions are good but apply equally to `main` :sweat_smile:. I think it would be better to include zpools we fail to access, and even on success we probably we ought to include more of what `get_info` returns (at least the zpool health status?). Mind if I file this as another separate issue?

Given all that, I'm not sure to do about the name, either. `collect_inventory` maybe?

---

Sorry for the noise - I can try to look at `main` first to make sure I'm not slowing down this work. fair enough.

---

I don't think it's noise. This is a legit deficiency; it just happens to not be introduced by this work. :) I still think it's worth lifting this to an issue of its own: #8184